### DISCUSSION TITLE
Option to include params of Annotated (pep593) types

### DISCUSSION THREAD

**Discussion by MajorDallas** [microsoft/pyright#4045] [Ideas]
I had the thought that it might be nice to be able to see the extra params given to `Annotated` in eg. editor hoverdocs. The current behavior is correct according to the PEP, so I don't propose any changes to what Pyright itself does with `Annotated` other than including the extra params in its output if the user opts-in to that behavior.

To borrow an example from [PEP 539](https://peps.python.org/pep-0593/):

```python
UnsignedShort = Annotated[int, struct2.ctype('H')]
SignedChar = Annotated[int, struct2.ctype('b')]

class Student(struct2.Packed):
    name: Annotated[str, struct2.ctype("<10s")]
    serialnum: UnsignedShort
    school: SignedChar

# Normal behavior:
reveal_type(Student.name)  # str

# With a "show Annotated" option enabled:
reveal_type(Student.name)  # Annotated[str, struct2.ctype("<10s")]
```

It shouldn't be necessary to resolve any expressions that might appear in the params list since the results of evaluating any such expression would only be meaningful to a runtime consumer and, ostensibly, a human reading it in a hoverdoc should already have some idea of what the expression would do at runtime. Just reporting the raw text should be sufficient.

Any thoughts? Is it even possible? (or god forbid, has this come up before and I missed it in my search before writing this?)

**Comment by erictraut** [comment]
When you hover over an expression with your mouse pointer or call `reveal_type`, pyright evaluates the type of the target expression _in its current context_. That involves a bunch of work including code flow analysis, type narrowing based on conditional type guards and assignments, etc. This process involves various type transforms — including narrowing, combining (unions), expanding (for overload matching), TypeVar solving, etc. These transforms are based on a well-defined type algebra rules.

The problem with `Annotated` is that it includes non-type metadata. A type checker has no way of knowing what that metadata means, how it should be interpreted, and how/whether it composes according to certain transforms, so there's really nothing that it can do with it. Even if it could evaluate this metadata and associated with a type, the first transform applied to that type would need to strip out that metadata because there is no defined algebra for this metadata. For example, what happens when two types are combined into a union? Does the metadata somehow get combined as well?

PEP 593 says that a type checker should ignore all arguments to `Annotated` other than the first. The second and subsequent arguments are value expressions, and a type checker doesn't know how to evaluate value expressions. It can attempt to infer the type of that expression (which is probably not what you care about). It doesn't know how to evaluate the expression's value.

For those reasons, I don't think it makes sense for a type checker or language server to do anything with an `Annotated` other than evaluate the first argument, which is assumed to be a valid type expression.
